### PR TITLE
Introduce /itchysats/order protocol over libp2p

### LINKED ISSUE #1770: Design proposal for `/itchysats/order`
This proposal is dependent on the scope extension mentioned in https://github.com/itchysats/itchysats/issues/1768.

Placing an order is a core abstraction in the financial industry. Placing an order in a system can result in it being filled at which point the positions of both contract partners are adjusted accordingly. Orders can also remain unfilled. In a marketplace, unfilled orders are gathered in an orderbook. They are the foundation for how the mid-market price is defined for example.

We can move one step closer to such a model through the following design:

- Peers can open a substream with the `/itchysats/order` protocol
- The dialing peer sends an `Order` structure to the listening peer. An `Order` struct contains (at least) the peer's intention (buy/sell), the asset, price, quantity and leverage.
- Because we are placing an order directly with a peer and for improved user experience, the listening peer SHOULD decide straight away whether the are willing to fill this order or not and communicate their response back to the dialer through an `Accept` or `Reject` message.

```mermaid
sequenceDiagram
    Taker->>Maker: Open substream `/itchysats/order`, send `Order` struct
    Maker->>Taker: Accept
```

## Interim solution

As an interm solution, we propose to directly attach the creation of a new position to the `/itchysats/order` protocol. We can change these semantics in a later version of the protocol.

```mermaid
sequenceDiagram
    Taker->>Maker: Open substream `/itchysats/order`, send `Order` struct
    Maker->>Taker: Accept
	Taker->>Taker: Transition to contract-setup
	Maker->>Maker: Transition to contract-setup
	Taker->>Maker: Msg1
	Maker->>Taker: Msg1
	Note over Taker,Maker: More messages ...
```

## Appendix: With combined position

At the moment, itchysats does not have a "combined position". Eventually, it would make sense to track a peer's position and adjust it based on the orders they are placing. Doing so would allow the `/itchysats/order` protocol to _end_ at this point and be followed by a separate protocol that - based on the new balance of each peer's account - aligns the blockchain state with the account state. This look something like this:

```mermaid
sequenceDiagram
	Note over Taker,Maker: Assume the taker is 500 contracts long
    Taker->>Maker: Open substream `/itchysats/order`, buy 200 contracts @ $54000 with x3 leverage
    Maker->>Taker: Accept
	Taker->>Taker: Adjust local position: +200 contracts
	Maker->>Maker: Adjust local position: +200 contracts
	Note over Taker,Maker: +200 contracts, requires adding more margin for maker and taker
	Taker->>Maker: Open substream `/itchysats/channel-update`
	Note over Taker,Maker: Exchange required messages for channel update
```

Decoupling the channel update from placing orders allows for more flexibility. It can for example be used to also adjust the leverage of the position without placing an order.

However, introducing these abstractions now is deemed to be too much of a fundamental change to how the software works.

### PR DESCRIPTION
**Pull request by klochowicz**
Fix https://github.com/itchysats/itchysats/issues/1770.

The new design follows the proposal in #1770. This means that
the taker now generates an `OrderId` when _placing_ the order. We
currently reuse the same type for the offer (legacy) and the taker's
order, which is a bit confusing (but hard to avoid, because our model
wants an `OrderId`). With the new protocol the taker ignores the
`OrderId` associated with the offer, but we have to keep it around for
legacy connections.

One actor test got removed - with libp2p we're not changing the offers after
one was picked; the OrderId is generated on the taker side.

Co-authored-by: Lucas Soriano del Pino <l.soriano.del.pino@gmail.com>

### CODE REVIEW

**Comment by klochowicz** [comment]
I have slightly modified @luckysori 's PR:
- removed `Contracts` changes, as they are orthogonal to the network migration, and the discussion on github is not resolved yet - moreover, they caused grief during rebasing, when types got migrated). 
- addressed some leftover TODOs
- made the tests pass 
- removed one actor level test (we no longer remove the offer after an order got taken)

I didn't end up going back to recreating the old protocol in order to save time; this solution was close to finish, and it fixes some awkwardness in our protocol (need to recreate offers). I would be keen on trying to let it run on testnet over the weekend.

File: daemon/src/order/maker.rs
Comment (da-kami) [inline daemon/src/order/maker.rs]: With the protocol change of letting the `taker` define the `OrderId` (independent from the `OfferId`), the maker has to make sure that *all* parameters provided by the taker are sane! This is not what is currently implemented.

I am happy to go that way, but I think this will require changes when handling the decision to accept/reject. If there is not ticket for that yet please create one.

File: daemon/src/order/taker.rs
Comment (da-kami) [inline daemon/src/order/taker.rs]:
```suggestion
                    maker_position,
```

Comment (da-kami) [reply daemon/src/order/taker.rs]: Otherwise the `position.counterposition()` below when creating the `Cfd` comes out of nowhere.

File: daemon/src/order/taker.rs
Comment (da-kami) [inline daemon/src/order/taker.rs:177]: I think it would be a good idea to split up contract setup into reusable chunks as well and then inline the messaging here. Can be done in a follow up.

**Review by da-kami — COMMENTED** [review]
I assume that you are aware that the protocol changes will require logic changes to the maker's decision to evaluate if an `order` is acceptable.

We should evaluate implementing at least some of the acceptance criteria before we can roll this out. Unless there is already a ticket for this (I did not find one) please create a ticket for that (automation). We have to make sure to do this before releasing.

File: daemon/src/order/taker.rs
Comment (klochowicz) [reply daemon/src/order/taker.rs:177]: I would only do it if it's necessary; if it works well without this, I see no point in doing so; I doubt that such a refactor would even be more maintainable...


> I assume that you are aware that the protocol changes will require logic changes to the maker's decision to evaluate if an `order` is acceptable.
> 
> We should evaluate implementing at least some of the acceptance criteria before we can roll this out. Unless you are already working on it please create a ticket for that. We have to make sure to do this before releasing.

great point. I will add it an additional commit here, because without this it's not a complete PR :)

**Comment by da-kami** [comment]
I like separating offer and order more clearly. 
We might also want to add tests that show that one taker (or different takers for that matter) can take the same `offer_id` twice now (resulting in different `order_ids`.

File: daemon/src/order/maker.rs
Comment (klochowicz) [reply daemon/src/order/maker.rs]: good point. will add a check :)

**Comment by klochowicz** [comment]
I have rebased the PR and finally made the tests pass again; it's not ready yet for another review though as I need to add:

TODO: 
- [x] re-add identity (we're using it in grafana now)
- [x] validate the offer by tying order and offer id together (as discussed with @da-kami )

Marking as draft until then.

**Comment by da-kami** [comment]
I rebased on latest master for now, will see to work in the changes of the `taker` sending the offer id when creating an order instead of all the parameters. 

@klochowicz what is the second TODO about? - i.e.:

>  re-add identity (we're using it in grafana now)

**Comment by klochowicz** [comment]
> I rebased on latest master for now, will see to work in the changes of the `taker` sending the offer id when creating an order instead of all the parameters.
> 
> @klochowicz what is the second TODO about? - i.e.:
> 
> > re-add identity (we're using it in grafana now)

the TODO was about re-adding the `identity` - @luckysori has hardcoded a random "identity" field as it wasn't used, but we're now using it for some reporting.

**Comment by da-kami** [comment]
> > I rebased on latest master for now, will see to work in the changes of the `taker` sending the offer id when creating an order instead of all the parameters.
> > @klochowicz what is the second TODO about? - i.e.:
> > > re-add identity (we're using it in grafana now)
> 
> the TODO was about re-adding the `identity` - @luckysori has hardcoded a random "identity" field as it wasn't used, but we're now using it for some reporting.

Sorry, but could you provide a bit more context? What "reporting" are we talking about? Is this important for the future as well? 
This identity is only relevant for the old connection as far as I understand, how would we do the "reporting" once we remove it?

**Comment by klochowicz** [comment]
> > > I rebased on latest master for now, will see to work in the changes of the `taker` sending the offer id when creating an order instead of all the parameters.
> > > @klochowicz what is the second TODO about? - i.e.:
> > > > re-add identity (we're using it in grafana now)
> > 
> > 
> > the TODO was about re-adding the `identity` - @luckysori has hardcoded a random "identity" field as it wasn't used, but we're now using it for some reporting.
> 
> Sorry, but could you provide a bit more context? What "reporting" are we talking about? Is this important for the future as well? This identity is only relevant for the old connection as far as I understand, how would we do the "reporting" once we remove it?

I've linked you the reports on matrix, and explained them in greater detail.

File: daemon/src/lib.rs
Comment (klochowicz) [inline daemon/src/lib.rs]: why is it hard to supervise this because of the tests? oracle actor is mocked in the tests, but we're using DI here, so it should not matter?

File: daemon/src/order/maker.rs
Comment (klochowicz) [inline daemon/src/order/maker.rs]: yes, it's expected that we can't use `send_async_next` here; we are awaiting the completion of the handler as we need the result back :)

note: we could use `send(msg).priority(NUMBER)` if we wanted to utilise higher priority to ensure that this protocol does go smoothly (at the risk of actually getting an outdated offer, so I'm not sure whether it's worth it)

File: daemon/src/order/maker.rs
Comment (klochowicz) [inline daemon/src/order/maker.rs:110]: in order to address the TODO below it would be good to instrument this function:

```suggestion
    #[instrument(skip(self), err)]
    async fn pick_offer(&self, offer_id: OfferId) -> Result<Offer> {
```

File: daemon/src/order/maker.rs
Comment (klochowicz) [inline daemon/src/order/maker.rs]:
```suggestion
            .context("Timeout when waiting for order")?
```

File: daemon/src/order/maker.rs
Comment (klochowicz) [inline daemon/src/order/maker.rs]:
```suggestion
    #[instrument(skip(self), err)]
    async fn await_order(
```

File: daemon/src/order/maker.rs
Comment (klochowicz) [inline daemon/src/order/maker.rs]: `message_impl` that used to implement message handlers is not needed anymore. :)

```suggestion
#[xtra_productivity]
```

File: daemon/src/order/maker.rs
Comment (klochowicz) [inline daemon/src/order/maker.rs:190]: oops, I was meant to reinstate this one! I only did it on the other side. we'll need to pass it in the message unfortunately, there's no other way. will add one more commit :)

File: daemon/src/order/maker.rs
Comment (klochowicz) [inline daemon/src/order/maker.rs]: nit: why lowercase?

File: daemon/src/order/taker.rs
Comment (klochowicz) [inline daemon/src/order/taker.rs:124]: do we need to return `Ok()` here? this obscures the error in the traces....

File: daemon/src/taker_cfd.rs
Comment (klochowicz) [inline daemon/src/taker_cfd.rs]: maybe at least add a TODO to rollout `OfferId` in daemon later, when we decide to drop legacy protocols?

File: daemon/src/taker_cfd.rs
Comment (klochowicz) [inline daemon/src/taker_cfd.rs]: 👍 I'm for removing this. maker should deal with outdated orders by just rejecting them. at least we have visibility how often it happens, otherwise we don't know that takers run into this frustrating problem.

File: maker/src/contract_setup.rs
Comment (klochowicz) [inline maker/src/contract_setup.rs]: should this be `offer`? 

```suggestion
   offer: Offer,
```

File: sqlite-db/src/rollover/insert.rs
Comment (klochowicz) [inline sqlite-db/src/rollover/insert.rs:84]: 🤔should this be order_id or offer_id? how can we tell which one of the two we need here?

File: xtra-libp2p-offer/src/lib.rs
Comment (klochowicz) [inline xtra-libp2p-offer/src/lib.rs]: 🙃

```suggestion
    fn dummy_offer(position: Position) -> Offer {
```

**Review by klochowicz — COMMENTED** [review]
great piece of work, thanks for linking OrderId to OfferId!

It seems I can't approve this PR, as I was the author of it? 😆

from my side ACK (after squashing before merging)

**Comment by klochowicz** [comment]
@da-kami @luckysori I've added in total 2 commits into this PR  that bring `Identity` (taker_id and maker_id) into Order protocol. please have a look at these too :)

File: daemon/src/lib.rs
Comment (luckysori) [reply daemon/src/lib.rs]: Unfortunately I don't remember the specifics because it has been so long since I wrote this comment. But I just couldn't make it work nicely. Feel free to give it a go!

File: daemon-tests/tests/offer.rs
Comment (luckysori) [inline daemon-tests/tests/offer.rs]: 🔧 Rename to `new_offer_id_{taker,maker}`.

File: daemon/src/order.rs
Comment (luckysori) [inline daemon/src/order.rs]: 🔧 I would expect this identifier to be:

- `/itchysats/order`; and
- version `1.0.0`, to be consistent with the first version of our other protocols.

File: daemon/src/order/maker.rs
Comment (luckysori) [inline daemon/src/order/maker.rs]: 🔧 Using `anyhow::Context` is good here.

```suggestion
        let latest_offers = match self
            .latest_offers
            .send(xtra_libp2p_offer::maker::GetLatestOffers)
            .await
            .context("Failed to retrieve latest offer from offers actor")?;
```

And the same goes for the two `match`es below.

File: daemon/src/order/maker.rs
Comment (luckysori) [inline daemon/src/order/maker.rs]: 🙃 Too many awaits in one line.

```suggestion
        let order = match self.receive_order(&mut framed).await {
```

File: daemon/src/order/maker.rs
Comment (luckysori) [inline daemon/src/order/maker.rs:180]: 🤔 Do we actually need this? If someone comes and tries to place an order, why does the `maker` daemon care about which offer they were referencing? Shouldn't the daemon just let the automation (or whoever is operating the application) decide, since they can?

I think this has been discussed before, but I keep forgetting the arguments because intuitively it doesn't make sense to me.

File: daemon/src/order/maker.rs
Comment (luckysori) [reply daemon/src/order/maker.rs]: This is definitely an area where we can have a problem. I can already imagine the offer actor, overwhelmed with it's usual job of relaying offers, taking ages to reply to this message.

File: daemon/src/order/taker.rs
Comment (luckysori) [reply daemon/src/order/taker.rs:124]: I don't see the alternative: the task is fallible and we are early returning here.

File: sqlite-db/src/rollover/insert.rs
Comment (da-kami) [reply sqlite-db/src/rollover/insert.rs:84]: This one was super confusing for myself, because when this was introduced we treated the offer as order (i.e. we always spoke of the offer as order and wanted to rename it to offer), hence @bonomat opted for naming this `offer_id`. However, with the introduction of clear boundaries for offer and order as part of my commit this actually is related to the order, and this *is* an `OrderId` here. From a lifecylce point of view when this is created the order is filled alrady.
Currently, the `offer_id` in the `Cfd` is just to identify the offer related to the `Cfd`, but not the `Cfd` itself. The `OrderId` identifies the `Cfd`, because if an order gets filled the taker creates the `Cfd` with a unique `OrderId`.


Note: The lifecycle of the `Cfd` can still be improved, we could introduce a dedicated `Order` and only once said `Order` is filled we create the actual `Cfd`. But I felt we can do that in another iteration on top of this work if we want to. So for now I kept the `cfds` table as a flattened out data conglomerate that includes all the information and the lifecylce start with creating the `Order` on the taker side. This means that, as before, we can have a `Cfd` in the database that was never actually filled by the maker - which is a bit weird. 
I think eventually a model of `Offer -> Order -> Cfd` would be correct, where the `offer_id` is linked to the `order_id` and the `order_id` *may be* the same as the `cfd_id`. One `offer_id` can be linked to multiple `order_id`s, one `order_id` should be linked to exactly one `cfd_id` (or `cfd_id` = `order_id`)

**Review by luckysori — COMMENTED** [review]
Haven't reviewed all of it, but I have to stop now until a bit later.

File: daemon/src/lib.rs
Comment (klochowicz) [reply daemon/src/lib.rs]: all good! I'll try after this gets in, was just curious :)

File: daemon/src/order/maker.rs
Comment (da-kami) [reply daemon/src/order/maker.rs]: 😅 I wonder what my brain was thinking when I wrote this

File: daemon/src/order/maker.rs
Comment (da-kami) [reply daemon/src/order/maker.rs:180]: I agree. This was discussed to some extent, but it would be a change that requires adapting the automation. I would like to keep this changeset contained and just keep the logic as it was before, and then change this in another iteration.

This is a change that purely concerns the maker, so we can change this easily on top, I will see to create a ticket.

Comment (da-kami) [reply daemon/src/order/maker.rs:180]: Note: I am happy to action this straight away, so we also solve the problem of not needing to know the latest offer in this actor. 
But, we have to evaluate how the automation then keeps the state of the current offer. Currently it just does not keep that state as far as I know, so I don't think this is such a simple change and would not see it in scope of this PR.

Comment (luckysori) [reply daemon/src/order/maker.rs:180]: Thanks for your reply! I think it's fine to not tackle it immediately, particularly if it's not a simple change. If you create a ticket and we tackle it soon I will be more than happy.

File: daemon/src/order/maker.rs
Comment (da-kami) [reply daemon/src/order/maker.rs]: I think we can "solve" this by moving this decision into the automation, but we have to be aware that we will have to manage the state there then. I will see to write down a ticket for this and action it later, because it is not just copying code ;)
For now I think this solution is good enough, and I don't expect to run into too much trouble, because sending to each peer is in a separate task, so the offer actor should not be blocked for too long.

File: daemon/src/order/maker.rs
Comment (da-kami) [reply daemon/src/order/maker.rs:180]: Created :) see linked tickets

File: sqlite-db/src/rollover/insert.rs
Comment (da-kami) [reply sqlite-db/src/rollover/insert.rs:84]: I created https://github.com/itchysats/itchysats/issues/2555 to shed some more light on this and outline how I envision to clean this up.

File: daemon/src/order/maker.rs
Comment (da-kami) [reply daemon/src/order/maker.rs]: I'll change it to the same name as the enum. @luckysori if this was important please let me know :)

File: daemon/src/taker_cfd.rs
Comment (da-kami) [reply daemon/src/taker_cfd.rs]: This should actually be `OfferId` here :)

File: daemon/src/taker_cfd.rs
Comment (da-kami) [reply daemon/src/taker_cfd.rs]: I disagree. The taker UI does not tell you that an offer is outdated, and I don't want to introduce this as part of this PR. If we want to remove this feature we can, but it should not be part if this changeset because it has notthing to do with it.
We introduces this to prevent the taker from taking outdated offers created by the maker (!) after we introduces replicating the offer in the daemon. This can still happen and this is a safety net for the taker.

File: maker/src/contract_setup.rs
Comment (da-kami) [reply maker/src/contract_setup.rs]: Yes and no 😅 
This is legacy code and I think no matter how much I try to change it, it will always be odd because the boundaries are just not right. I agree that there is quite a lot of weirdness in this file because of renaming `Order` to `Offer`. I am not sure if it's a good idea to try to clean this up if we remove it soon.

The problem is: If I change it here then outside it will be super weird, because we are passing the `order_to_take` in, and that cannot really be changed, because it is a construct that is offer and order at the same time 😅

Comment (da-kami) [reply maker/src/contract_setup.rs:30]: Given that this is the only instance where we literally have a direct conflict I'll change that one :)

**Comment by da-kami** [comment]
Interesting: installing an using a new `sqlx-cli` to prepare the db can indeed lead to the changes in the `sqlx-data.json` file. I just had such an instance, where the CI complained and I could only reproduce it after installing the new version.

File: daemon/src/order/maker.rs
Comment (da-kami) [inline daemon/src/order/maker.rs:197]: Note: We will be able to overcome this with #2555

**Review by da-kami — APPROVED** [review]

**Pull request merged**
